### PR TITLE
Create symlinks at runtime and don't use relative paths

### LINKED ISSUE #3598: 0.9.1 test failures related to test certificates and symlinks
I'm working on updating the GNU Guix packaging of certbot [0] to 0.9.1.

The tests 'test_ancient_webroot_renewal_conf', 'test_quiet_renew', 'test_renew_no_hook_validation', and 'test_renew_verb' fail (detailed output below).

The primary cause of the failure seems to be that the test certs are expected to be symlinks, but they are not:
```
CertStorageError: expected /tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/testdata/live/sample-renewal/cert.pem to be a symlink
```

I noticed that the Arch Linux package manually makes these files into symlinks before building [1], and the tests do pass for me when I make a similar change in our packaging.

What do you recommend we do?

[0]
http://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/tls.scm#n524

[1]
https://git.archlinux.org/svntogit/community.git/commit/trunk?h=packages/certbot&id=6d8c2e89be969ba73148a6302c17059a6ccfd93d

Detailed output from test suite:

```
======================================================================
ERROR: test_ancient_webroot_renewal_conf (certbot.tests.cli_test.CLITest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/gnu/store/vyjhybsw6zpq27c55bb737zwq65kn7by-python2-mock-1.0.1/lib/python2.7/site-packages/mock.py", line 1201, in patched
    return func(*args, **keywargs)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 710, in test_ancient_webroot_renewal_conf
    configuration.RenewerConfiguration(config))
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/storage.py", line 259, in __init__
    self._check_symlinks()
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/storage.py", line 267, in _check_symlinks
    "expected {0} to be a symlink".format(link))
CertStorageError: expected /tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/testdata/live/sample-renewal/cert.pem to be a symlink

======================================================================
FAIL: test_quiet_renew (certbot.tests.cli_test.CLITest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 681, in test_quiet_renew
    _, _, stdout = self._test_renewal_common(True, [], args=args, should_renew=True)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 615, in _test_renewal_common
    traceback.format_exc())
AssertionError: Unexpected renewal error:
Traceback (most recent call last):
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 606, in _test_renewal_common
    ret, stdout, _, _ = self._call(args)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 64, in _call
    ret, stdout, stderr = self._call_no_clientmock(args, stdout)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 74, in _call_no_clientmock
    ret = main.main(args[:])  # NOTE: parser can alter its args!
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/main.py", line 776, in main
    return config.func(config, plugins)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/main.py", line 592, in renew
    renewal.renew_all_lineages(config)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/renewal.py", line 365, in renew_all_lineages
    len(renew_failures), len(parse_failures)))
Error: 0 renew failure(s), 1 parse failure(s)


======================================================================
FAIL: test_renew_no_hook_validation (certbot.tests.cli_test.CLITest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 701, in test_renew_no_hook_validation
    error_expected=False)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 615, in _test_renewal_common
    traceback.format_exc())
AssertionError: Unexpected renewal error:
Traceback (most recent call last):
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 606, in _test_renewal_common
    ret, stdout, _, _ = self._call(args)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 64, in _call
    ret, stdout, stderr = self._call_no_clientmock(args, stdout)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 74, in _call_no_clientmock
    ret = main.main(args[:])  # NOTE: parser can alter its args!
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/main.py", line 776, in main
    return config.func(config, plugins)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/main.py", line 592, in renew
    renewal.renew_all_lineages(config)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/renewal.py", line 365, in renew_all_lineages
    len(renew_failures), len(parse_failures)))
Error: 0 renew failure(s), 1 parse failure(s)


======================================================================
FAIL: test_renew_verb (certbot.tests.cli_test.CLITest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 676, in test_renew_verb
    self._test_renewal_common(True, [], args=args, should_renew=True)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 615, in _test_renewal_common
    traceback.format_exc())
AssertionError: Unexpected renewal error:
Traceback (most recent call last):
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 606, in _test_renewal_common
    ret, stdout, _, _ = self._call(args)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 64, in _call
    ret, stdout, stderr = self._call_no_clientmock(args, stdout)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/cli_test.py", line 74, in _call_no_clientmock
    ret = main.main(args[:])  # NOTE: parser can alter its args!
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/main.py", line 776, in main
    return config.func(config, plugins)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/main.py", line 592, in renew
    renewal.renew_all_lineages(config)
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/renewal.py", line 365, in renew_all_lineages
    len(renew_failures), len(parse_failures)))
Error: 0 renew failure(s), 1 parse failure(s)


----------------------------------------------------------------------
Ran 517 tests in 76.819s

FAILED (failures=3, errors=1)
Saving debug log to /tmp/guix-build-certbot-0.9.1.drv-0/tmpiL5LZK/logs/letsencrypt.log
2016-10-07 06:30:25,544:DEBUG:certbot.main:certbot version: 0.9.1
2016-10-07 06:30:25,546:DEBUG:certbot.main:Arguments: ['--config-dir', '/tmp/guix-build-certbot-0.9.1.drv-0/tmpiL5LZK/config', '--work-dir', '/tmp/guix-build-certbot-0.9.1.drv-0/tmpiL5LZK/work', '--logs-dir', '/tmp/guix-build-certbot-0.9.1.drv-0/tmpiL5LZK/logs', '--text', '--dry-run', '-tvv']
2016-10-07 06:30:25,549:DEBUG:certbot.main:Discovered plugins: PluginsRegistry(PluginEntryPoint#webroot,PluginEntryPoint#null,PluginEntryPoint#manual,PluginEntryPoint#standalone)
2016-10-07 06:30:25,555:WARNING:certbot.renewal:expected /tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/testdata/live/sample-renewal/cert.pem to be a symlink
2016-10-07 06:30:25,566:WARNING:certbot.renewal:Renewal configuration file /tmp/guix-build-certbot-0.9.1.drv-0/tmpiL5LZK/config/renewal/sample-renewal.conf is broken. Skipping.
2016-10-07 06:30:25,576:DEBUG:certbot.renewal:Traceback was:
Traceback (most recent call last):
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/renewal.py", line 62, in _reconstitute
    full_path, configuration.RenewerConfiguration(config))
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/storage.py", line 259, in __init__
    self._check_symlinks()
  File "/tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/storage.py", line 267, in _check_symlinks
    "expected {0} to be a symlink".format(link))
CertStorageError: expected /tmp/guix-build-certbot-0.9.1.drv-0/certbot-0.9.1/certbot/tests/testdata/live/sample-renewal/cert.pem to be a symlink


Please deploy a DNS TXT record under the name
_acme-challenge.foo.com with the following value:

Slui_IKoxhFjRRete7mBO4dnCu24lJt4G4q1k-CO1ho

Once this is deployed,
IMPORTANT NOTES:
 - If you like Certbot, please consider supporting our work by:

   Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
   Donating to EFF:                    https://eff.org/donate-le

IMPORTANT NOTES:
 - If you like Certbot, please consider supporting our work by:

   Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
   Donating to EFF:                    https://eff.org/donate-le
```

### PR DESCRIPTION
**Pull request by bmw**
Fixes #3598.

### CODE REVIEW

File: certbot/tests/cli_test.py [hunk 1]
```diff
@@ -1,4 +1,7 @@
 """Tests for certbot.cli."""
+# Many tests in this file should be moved
+# into main_test.py and renewal_test.py.
```
Comment (cowlicks) [inline certbot/tests/cli_test.py]: Make an issue?

File: certbot/tests/cli_test.py [hunk 2]
```diff
@@ -578,7 +578,7 @@ def test_certonly_new_request_failure(self):
     def _test_renewal_common(self, due_for_renewal, extra_args, log_out=None,
                              args=None, should_renew=True, error_expected=False):
         # pylint: disable=too-many-locals,too-many-arguments
-        cert_path = 'certbot/tests/testdata/cert.pem'
+        cert_path = test_util.vector_path('cert.pem')
```
Comment (cowlicks) [inline certbot/tests/cli_test.py]: Any idea why it is called *vector* path?

File: certbot/tests/cli_test.py (hunk 1 above)
Comment (bmw) [inline certbot/tests/cli_test.py]: It's #2716, but I should reference that here. I'll send a follow up commit.

Nice catch though. In general, I don't like all the `TODO` and `XXX` comments in our code base. I think if an actual issue is made, it's much more likely to get done.

Also, for context, what happened is once upon a time, `cli.py` was an absolutely massive file. That file was split into `cli.py,` `main.py`, and `renewal.py`, but the tests weren't split into separate files as well.

File: certbot/tests/cli_test.py (hunk 2 above)
Comment (bmw) [inline certbot/tests/cli_test.py]: I believe it refers to [test vector](https://en.wikipedia.org/wiki/Test_vector).

File: certbot/tests/cli_test.py (hunk 1 above)
Comment (cowlicks) [inline certbot/tests/cli_test.py]: Ah I see. I noticed this problem while working on the tests for account deactivation.

**Pull request merged**